drivers/ethernet/dwc_xgmac/eth_dwc_xgmac.c

```diff
@@ -0,0 +1,1943 @@
+	eth_enable_irq_t irq_enable_fn;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
+	/* XGMAC peripheral reset signal */
+	const struct reset_dt_spec reset;
+#endif
+};
+
+static inline mem_addr_t get_reg_base_addr(const struct device *dev)
+{
+	return (mem_addr_t)DEVICE_MMIO_GET(dev);
+}
+
+static inline int dwxgmac_software_reset(mem_addr_t ioaddr)
+{
+	uint32_t delay_us = CONFIG_ETH_DWC_XGMAC_SW_RESET_TIMEOUT;
+	bool ret;
+	/* software reset */
+	mem_addr_t reg_addr = (mem_addr_t)(ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_MODE_OFST);
+
+	sys_write32(DMA_MODE_SWR_SET(1u), reg_addr);
+
+	ret = WAIT_FOR(!(sys_read32(reg_addr) & DMA_MODE_SWR_SET_MSK), delay_us, k_msleep(1));
+	if (ret == false) {
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
+static void dwxgmac_dma_init(const struct device *dev, const struct xgmac_dma_cfg *const dma_cfg)
+{
+	mem_addr_t ioaddr = get_reg_base_addr(dev);
+	mem_addr_t reg_addr =
+		(mem_addr_t)(ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_SYSBUS_MODE_OFST);
+
+	uint32_t reg_val = DMA_SYSBUS_MODE_RD_OSR_LMT_SET(dma_cfg->rd_osr_lmt) |
+			   DMA_SYSBUS_MODE_WR_OSR_LMT_SET(dma_cfg->wr_osr_lmt) |
+			   DMA_SYSBUS_MODE_AAL_SET(dma_cfg->aal) |
+			   DMA_SYSBUS_MODE_EAME_SET(dma_cfg->eame) |
+			   DMA_SYSBUS_MODE_BLEN4_SET(dma_cfg->blen4) |
+			   DMA_SYSBUS_MODE_BLEN8_SET(dma_cfg->blen8) |
+			   DMA_SYSBUS_MODE_BLEN16_SET(dma_cfg->blen16) |
+			   DMA_SYSBUS_MODE_BLEN32_SET(dma_cfg->blen32) |
+			   DMA_SYSBUS_MODE_BLEN64_SET(dma_cfg->blen64) |
+			   DMA_SYSBUS_MODE_BLEN128_SET(dma_cfg->blen128) |
+			   DMA_SYSBUS_MODE_BLEN256_SET(dma_cfg->blen256) |
+			   DMA_SYSBUS_MODE_UNDEF_SET(dma_cfg->ubl);
+
+	sys_write32(reg_val, reg_addr);
+
+	reg_addr = (ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_TX_EDMA_CONTROL_OFST);
+
+	reg_val = DMA_TX_EDMA_CONTROL_TDPS_SET(dma_cfg->edma_tdps);
+
+	sys_write32(reg_val, reg_addr);
+
+	reg_addr = (ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_RX_EDMA_CONTROL_OFST);
+
+	reg_val = DMA_RX_EDMA_CONTROL_RDPS_SET(dma_cfg->edma_rdps);
+
+	sys_write32(reg_val, reg_addr);
+	LOG_DBG("%s: DMA engine common initialization completed", dev->name);
+}
+
+static void dwxgmac_dma_chnl_init(const struct device *dev,
+				  const struct eth_dwc_xgmac_config *const config,
+				  struct eth_dwc_xgmac_dev_data *const data)
+{
+	uint32_t dma_chnl;
+	uint32_t max_dma_chnl = config->num_dma_chnl;
+	struct xgmac_dma_chnl_config *const dma_chnl_cfg =
+		(struct xgmac_dma_chnl_config *)&config->dma_chnl_cfg;
+	struct xgmac_dma_tx_desc_meta *tx_desc_meta;
+	struct xgmac_dma_rx_desc_meta *rx_desc_meta;
+	uint32_t reg_val;
+	mem_addr_t ioaddr = get_reg_base_addr(dev);
+	mem_addr_t reg_addr;
+
+	for (dma_chnl = 0; dma_chnl < max_dma_chnl; dma_chnl++) {
+		tx_desc_meta = (struct xgmac_dma_tx_desc_meta *)&data->tx_desc_meta[dma_chnl];
+		rx_desc_meta = (struct xgmac_dma_rx_desc_meta *)&data->rx_desc_meta[dma_chnl];
+
+		reg_addr = (mem_addr_t)(ioaddr + XGMAC_DMA_CHNLx_BASE_ADDR_OFFSET(dma_chnl) +
```

Comment:
why the mem_addr_t cast here if you do not do it afterwards? (so it's superfluous then?)

Reply:
Removed this type of casting. this one got missed during the code cleanup.